### ISSUE TITLE
rule proposal: compat/special-param-mutation

### ISSUE THREAD

**Issue by ss-o** [z-shell/zsh-lint#158]
Parent: #72

## Rule ID and name

Proposed `compat/special-param-mutation`, Mutating shell-set parameters from function scope.

## Proposed behavior

Report an assignment-only statement inside function scope when Zsh dynamic binding resolves it to a curated shell-set outer parameter rather than to a local binding declared in the current or an enclosing function.

## Bad and good examples

```zsh
# Bad: mutates the shell-set outer binding.
probe() {
  ZSH_VERSION=$1
}

# Good: purpose-specific local state.
probe() {
  local target_zsh_version=$1
}
```

## Severity

Warning. The assignment can alter shell-managed state beyond the function return and change what nested code observes.

## Evidence review

The current six-repository, 20-file reference corpus and full local z-shell organization search contain no qualifying assignment-only mutation of the curated parameters. Broader GitHub code search found temporary command-prefix assignments such as `ZSH_VERSION= source ...`; those are distinct, intentional command environments and must not be reported as persistent bare mutations. No qualifying production example was established.

The proposal therefore does not meet the repository rule policy's real-code evidence gate.

## False-positive and analysis boundary

An implementation would need to distinguish:

- assignment-only statements from command-prefix environment assignments;
- a prior local declaration in the current function;
- a dynamically visible local declaration in an enclosing caller or lexically nested function;
- assignment order and declarations that occur only after the assignment;
- intentional top-level configuration and test-harness shims.

The current `internal/scope` index stores only one symbol per name, does not model dominance or containing-function assignment order, and cannot prove the target binding. Extending it is an architecture change, not a small addition to `compat/special-param-shadow`.

## Parser dependency

The AST can represent assignment statements, but correct resolution needs an ordered dynamic-scope model and focused native Zsh probes. Native Zsh remains semantic authority.

## Suppression

If admitted later, use the shared directive with a reason:

```zsh
# zsh-lint disable=compat/special-param-mutation -- intentional compatibility shim
```

## Priority and effort

Priority P3, effort L because the scope model must be designed and validated before the rule.

## Disposition

Not planned because qualifying real-code evidence is absent and the required dynamic-binding model has no other admitted consumer. Reopen only with production evidence and a focused scope-index design justified independently of this single rule.
